### PR TITLE
Improve type annotations for `core.(trajectory/units)`

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

- Improve type annotations for `core.trajectory` and `core.units`
- Nest `_get_si_unit` and `_check_mappings` inside the only methods that call them in `Unit`

### Extend the following to support argument "unit" as both `str` and `Unit`
- `__init/new__` of `FloatWithUnit` and `ArrayWithUnit`
- `get_conversion_factor` of `Unit`, and `to` methods of `FloatWithUnit` and `ArrayWithUnit`

The type hint suggests `str`, docstring suggest `Unit` and implementation seems to handle **partly** of both.
https://github.com/materialsproject/pymatgen/blob/bb68c780834fea588f35420e1835ed3b21894a44/pymatgen/core/units.py#L295-L305
Might be good to support both for more flexibility, without sacrificing backwards compatibility.

#### TODO in future PR:
- Types for `core.structure`

### CODE REVIEW

File: pymatgen/core/units.py [hunk 1]
```diff
@@ -435,22 +444,25 @@ def as_base_units(self):
...
+        if self.unit_type is None:
+            raise RuntimeError("Cannot get supported unit for None.")
+
+        return tuple(ALL_UNITS[self.unit_type])
 
 
 class ArrayWithUnit(np.ndarray):
     """Subclasses numpy.ndarray to attach a unit type. Typically, you should
     use the pre-defined unit type subclasses such as EnergyArray,
     LengthArray, etc. instead of using ArrayWithFloatWithUnit directly.
```
Comment (DanielYang59) [inline pymatgen/core/units.py:463]: Not sure what is `ArrayWithFloatWithUnit`

**Comment by DanielYang59** [comment]
Can I have this one reviewed please? @janosh

**Review by janosh — APPROVED** [review]
thanks @DanielYang59! 👍

**Comment by DanielYang59** [comment]
Thanks for reviewing! Appreciate that.

**Pull request merged**